Snapshoting of machines failed if snapshot made second time for given machine

**Reproduction Steps:**

1. Create new machine.
2. Snapshot machine.
3. Stop machine.
4. Start machine from saved snapshot.
5. Attempt to snapshot machine.

**Expected behavior:**

Snapshot is made overwriting old one, possible display confirmation dialog to user.

**Observed behavior:**

Error shown:
```
2016-09-05 14:30:13,003[aceManager-3600]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 745]    - Snapshot with id 'snapshot******' or combination of workspace id 'workspace*****' & environment 'default' & machine name  'default'  already exists
org.eclipse.che.api.machine.server.exception.SnapshotException: Snapshot with id 'snapshot*****' or combination of workspace id 'workspace******' & environment 'default' & machine name  'default'  already exists
	at com.codenvy.api.dao.mongo.SnapshotDaoImpl.saveSnapshot(SnapshotDaoImpl.java:92) ~[codenvy-hosted-platform-api-impl-4.7.0.jar:4.7.0]
	at org.eclipse.che.api.workspace.server.WorkspaceManager.createSnapshotSync(WorkspaceManager.java:740) [che-core-api-workspace-4.7.0.jar:4.7.0]
	at org.eclipse.che.api.workspace.server.WorkspaceManager.lambda$createSnapshot$0(WorkspaceManager.java:518) [che-core-api-workspace-4.7.0.jar:4.7.0]
	at org.eclipse.che.api.workspace.server.WorkspaceManager$$Lambda$238/772546380.run(Unknown Source) [che-core-api-workspace-4.7.0.jar:4.7.0]
	at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) [che-core-commons-lang-4.7.0.jar:4.7.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
```
**Che version:**    4.7
**OS and version:** Centos 7.2
**Docker version:** 1.11.1
**Che install:**  Codenvy 4.7

**Additional information:**
**_This problem is reproduced only hosted version CHE_**

--- a/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineConfig.java
+++ b/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineConfig.java
@@ -18,7 +18,6 @@ import java.util.Map;
 /**
  * @author gazarenkov
  */
-@Deprecated
 public interface MachineConfig {
 
     /**

--- a/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineLimits.java
+++ b/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineLimits.java
@@ -15,7 +15,6 @@ package org.eclipse.che.api.core.model.machine;
  *
  * @author Alexander Garagatyi
  */
-@Deprecated
 public interface MachineLimits {
     /** Get memory size (in megabytes) that is allocated for starting machine. */
     int getRam();

--- a/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineSource.java
+++ b/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/MachineSource.java
@@ -13,7 +13,6 @@ package org.eclipse.che.api.core.model.machine;
 /**
  * @author gazarenkov
  */
-@Deprecated
 public interface MachineSource {
 
     /**

--- a/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/ServerConf.java
+++ b/core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/machine/ServerConf.java
@@ -17,7 +17,6 @@ import org.eclipse.che.commons.annotation.Nullable;
  *
  * @author Alexander Garagatyi
  */
-@Deprecated
 public interface ServerConf {
     /**
      * Reference to this server.

--- a/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerMachineSource.java
+++ b/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerMachineSource.java
@@ -25,7 +25,6 @@ import static org.eclipse.che.plugin.docker.machine.DockerInstanceProvider.DOCKE
  *
  * @author Florent Benoit
  */
-@Deprecated
 public class DockerMachineSource extends MachineSourceImpl {
 
     /**

--- a/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java
+++ b/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java
@@ -48,8 +48,7 @@ public class DockerExtConfBindingProvider implements Provider<String> {
         }
         File extConfDir = new File(localConfDir, PLUGIN_CONF);
         if (!extConfDir.isDirectory()) {
-            LOG.warn("DockerExtConfBindingProvider",
-                     String.format("%s set to the %s but it must be directory not file", CheBootstrap.CHE_LOCAL_CONF_DIR, localConfDir));
+            LOG.warn("{} set to the {} but it must be directory not file", CheBootstrap.CHE_LOCAL_CONF_DIR, localConfDir);
             return null;
         }
 

--- a/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineConfigDto.java
+++ b/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineConfigDto.java
@@ -25,7 +25,6 @@ import static org.eclipse.che.api.core.factory.FactoryParameter.Obligation.OPTIO
 /**
  * @author Alexander Garagatyi
  */
-@Deprecated
 @DTO
 public interface MachineConfigDto extends MachineConfig, Hyperlinks {
     @Override

--- a/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineLimitsDto.java
+++ b/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineLimitsDto.java
@@ -16,7 +16,6 @@ import org.eclipse.che.dto.shared.DTO;
 /**
  * @author Alexander Garagatyi
  */
-@Deprecated
 @DTO
 public interface MachineLimitsDto extends MachineLimits {
     MachineLimitsDto withRam(int memorySizeMB);

--- a/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineSourceDto.java
+++ b/wsmaster/che-core-api-machine-shared/src/main/java/org/eclipse/che/api/machine/shared/dto/MachineSourceDto.java
@@ -20,7 +20,6 @@ import static org.eclipse.che.api.core.factory.FactoryParameter.Obligation.OPTIO
 /**
  * @author Alexander Garagatyi
  */
-@Deprecated
 @DTO
 public interface MachineSourceDto extends MachineSource {
     @Override

--- a/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineConfigImpl.java
+++ b/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineConfigImpl.java
@@ -27,7 +27,6 @@ import java.util.stream.Collectors;
  *
  * @author Eugene Voevodin
  */
-@Deprecated
 public class MachineConfigImpl implements MachineConfig {
 
     public static MachineConfigImplBuilder builder() {

--- a/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineLimitsImpl.java
+++ b/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineLimitsImpl.java
@@ -15,7 +15,6 @@ import org.eclipse.che.api.core.model.machine.MachineLimits;
 /**
  * @author Alexander Garagatyi
  */
-@Deprecated
 public class MachineLimitsImpl implements MachineLimits {
     private final int memory;
 

--- a/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineSourceImpl.java
+++ b/wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/MachineSourceImpl.java
@@ -23,7 +23,6 @@ import static java.util.Objects.hash;
  *
  * @author Eugene Voevodin
  */
-@Deprecated
 public class MachineSourceImpl implements MachineSource {
 
     private String type;

--- a/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java
+++ b/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java
@@ -469,17 +469,15 @@ public class CheEnvironmentEngine {
      *
      * @param snapshot
      *         description of snapshot that should be removed
+     * @throws NotFoundException
+     *         if snapshot is not found
      * @throws ServerException
      *         if error occurs on snapshot removal
      */
-    public void removeSnapshot(SnapshotImpl snapshot) throws ServerException {
+    public void removeSnapshot(SnapshotImpl snapshot) throws ServerException, NotFoundException {
         final String instanceType = snapshot.getType();
-        try {
-            final InstanceProvider instanceProvider = machineInstanceProviders.getProvider(instanceType);
-            instanceProvider.removeInstanceSnapshot(snapshot.getMachineSource());
-        } catch (NotFoundException e) {
-            throw new ServerException(e.getLocalizedMessage(), e);
-        }
+        final InstanceProvider instanceProvider = machineInstanceProviders.getProvider(instanceType);
+        instanceProvider.removeInstanceSnapshot(snapshot.getMachineSource());
     }
 
     private void initializeEnvironment(String workspaceId,

--- a/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
+++ b/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
@@ -49,6 +49,7 @@ import javax.inject.Named;
 import javax.inject.Singleton;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
@@ -732,16 +733,9 @@ public class WorkspaceManager {
                                      .withWorkspaceId(workspaceId));
         String devMachineSnapshotFailMessage = null;
         for (MachineImpl machine : runtime.getMachines()) {
-            try {
-                SnapshotImpl snapshot = runtimes.saveMachine(namespace,
-                                                             workspaceId,
-                                                             machine.getId());
-                snapshotDao.saveSnapshot(snapshot);
-            } catch (ApiException apiEx) {
-                if (machine.getConfig().isDev()) {
-                    devMachineSnapshotFailMessage = apiEx.getLocalizedMessage();
-                }
-                LOG.error(apiEx.getLocalizedMessage(), apiEx);
+            Optional<String> error = createSnapshot(machine, namespace);
+            if (error.isPresent() && machine.getConfig().isDev()) {
+                devMachineSnapshotFailMessage = error.get();
             }
         }
         if (devMachineSnapshotFailMessage != null) {
@@ -755,6 +749,31 @@ public class WorkspaceManager {
                                          .withWorkspaceId(workspaceId));
         }
         return devMachineSnapshotFailMessage == null;
+    }
+
+    private Optional<String> createSnapshot(MachineImpl machine, String namespace) {
+        try {
+            try {
+                SnapshotImpl oldSnapshot = snapshotDao.getSnapshot(machine.getWorkspaceId(),
+                                                                   machine.getEnvName(),
+                                                                   machine.getConfig().getName());
+                snapshotDao.removeSnapshot(oldSnapshot.getId());
+
+                runtimes.removeSnapshot(oldSnapshot);
+            } catch (NotFoundException ignored) {
+                // Do nothing if no snapshot found
+            }
+
+            SnapshotImpl snapshot = runtimes.saveMachine(namespace,
+                                                         machine.getWorkspaceId(),
+                                                         machine.getId());
+            snapshotDao.saveSnapshot(snapshot);
+
+            return Optional.empty();
+        } catch (ApiException apiEx) {
+            LOG.error(apiEx.getLocalizedMessage(), apiEx);
+            return Optional.of(apiEx.getLocalizedMessage());
+        }
     }
 
     @VisibleForTesting

--- a/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
+++ b/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
@@ -764,10 +764,23 @@ public class WorkspaceManager {
                 // Do nothing if no snapshot found
             }
 
-            SnapshotImpl snapshot = runtimes.saveMachine(namespace,
-                                                         machine.getWorkspaceId(),
-                                                         machine.getId());
-            snapshotDao.saveSnapshot(snapshot);
+            SnapshotImpl snapshot = null;
+            try {
+                snapshot = runtimes.saveMachine(namespace,
+                                                machine.getWorkspaceId(),
+                                                machine.getId());
+
+                snapshotDao.saveSnapshot(snapshot);
+            } catch (ApiException e) {
+                if (snapshot != null) {
+                    try {
+                        runtimes.removeSnapshot(snapshot);
+                    } catch (ApiException e1) {
+                        LOG.error(e1.getLocalizedMessage(), e1);
+                    }
+                }
+                throw e;
+            }
 
             return Optional.empty();
         } catch (ApiException apiEx) {

--- a/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
+++ b/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java
@@ -49,7 +49,6 @@ import javax.inject.Named;
 import javax.inject.Singleton;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
@@ -733,9 +732,9 @@ public class WorkspaceManager {
                                      .withWorkspaceId(workspaceId));
         String devMachineSnapshotFailMessage = null;
         for (MachineImpl machine : runtime.getMachines()) {
-            Optional<String> error = createSnapshot(machine, namespace);
-            if (error.isPresent() && machine.getConfig().isDev()) {
-                devMachineSnapshotFailMessage = error.get();
+            String error = replaceSnapshot(machine, namespace);
+            if (error != null && machine.getConfig().isDev()) {
+                devMachineSnapshotFailMessage = error;
             }
         }
         if (devMachineSnapshotFailMessage != null) {
@@ -751,7 +750,7 @@ public class WorkspaceManager {
         return devMachineSnapshotFailMessage == null;
     }
 
-    private Optional<String> createSnapshot(MachineImpl machine, String namespace) {
+    private String replaceSnapshot(MachineImpl machine, String namespace) {
         try {
             try {
                 SnapshotImpl oldSnapshot = snapshotDao.getSnapshot(machine.getWorkspaceId(),
@@ -776,16 +775,19 @@ public class WorkspaceManager {
                     try {
                         runtimes.removeSnapshot(snapshot);
                     } catch (ApiException e1) {
-                        LOG.error(e1.getLocalizedMessage(), e1);
+                        LOG.error(format("Snapshot removal failed. Snapshot: %s. Error: %s",
+                                         snapshot,
+                                         e1.getLocalizedMessage()),
+                                  e1);
                     }
                 }
                 throw e;
             }
 
-            return Optional.empty();
+            return null;
         } catch (ApiException apiEx) {
-            LOG.error(apiEx.getLocalizedMessage(), apiEx);
-            return Optional.of(apiEx.getLocalizedMessage());
+            LOG.error("Snapshot creation failed. Error: " + apiEx.getLocalizedMessage(), apiEx);
+            return apiEx.getLocalizedMessage();
         }
     }
 

--- a/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java
+++ b/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java
@@ -490,10 +490,12 @@ public class WorkspaceRuntimes {
      *
      * @param snapshot
      *         description of snapshot that should be removed
+     * @throws NotFoundException
+     *         if snapshot is not found
      * @throws ServerException
      *         if error occurs
      */
-    public void removeSnapshot(SnapshotImpl snapshot) throws ServerException {
+    public void removeSnapshot(SnapshotImpl snapshot) throws ServerException, NotFoundException {
         environmentEngine.removeSnapshot(snapshot);
     }
 

--- a/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
+++ b/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
@@ -12,6 +12,7 @@ package org.eclipse.che.api.workspace.server;
 
 import org.eclipse.che.api.core.ConflictException;
 import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
 import org.eclipse.che.api.core.model.machine.MachineStatus;
 import org.eclipse.che.api.core.model.workspace.WorkspaceConfig;
 import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
@@ -34,6 +35,7 @@ import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
 import org.eclipse.che.api.workspace.server.model.impl.WorkspaceRuntimeImpl;
 import org.eclipse.che.api.workspace.server.spi.WorkspaceDao;
 import org.eclipse.che.api.workspace.shared.Constants;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
 import org.eclipse.che.commons.env.EnvironmentContext;
 import org.eclipse.che.commons.subject.Subject;
 import org.eclipse.che.commons.subject.SubjectImpl;
@@ -48,6 +50,7 @@ import org.testng.annotations.Test;
 
 import java.util.HashMap;
 import java.util.List;
+import java.util.UUID;
 
 import static com.google.common.base.Strings.isNullOrEmpty;
 import static java.util.Arrays.asList;
@@ -59,6 +62,10 @@ import static org.eclipse.che.api.core.model.workspace.WorkspaceStatus.STOPPED;
 import static org.eclipse.che.api.workspace.server.WorkspaceManager.CREATED_ATTRIBUTE_NAME;
 import static org.eclipse.che.api.workspace.server.WorkspaceManager.UPDATED_ATTRIBUTE_NAME;
 import static org.eclipse.che.api.workspace.shared.Constants.AUTO_RESTORE_FROM_SNAPSHOT;
+import static org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent.EventType.SNAPSHOT_CREATED;
+import static org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent.EventType.SNAPSHOT_CREATING;
+import static org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent.EventType.SNAPSHOT_CREATION_ERROR;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
 import static org.mockito.Matchers.any;
 import static org.mockito.Matchers.anyBoolean;
 import static org.mockito.Matchers.anyObject;
@@ -527,13 +534,16 @@ public class WorkspaceManagerTest {
 
     @Test
     public void shouldBeAbleToStopWorkspace() throws Exception {
+        // given
         final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
         final RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
         when(runtimes.get(any())).thenReturn(descriptor);
         when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
 
+        // when
         workspaceManager.stopWorkspace(workspace.getId());
 
+        // then
         verify(runtimes, timeout(2000)).stop(workspace.getId());
         assertNotNull(workspace.getAttributes().get(UPDATED_ATTRIBUTE_NAME));
     }
@@ -545,6 +555,11 @@ public class WorkspaceManagerTest {
         when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
         final RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
         when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
 
         workspaceManager.stopWorkspace(workspace.getId());
 
@@ -618,6 +633,7 @@ public class WorkspaceManagerTest {
 
     @Test
     public void shouldCreateWorkspaceSnapshotUsingDefaultValueForAutoRestore() throws Exception {
+        // given
         workspaceManager = spy(new WorkspaceManager(workspaceDao,
                                                     runtimes,
                                                     eventService,
@@ -628,9 +644,16 @@ public class WorkspaceManagerTest {
         when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
         final RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
         when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
 
+        // then
         workspaceManager.stopWorkspace(workspace.getId());
 
+        // then
         verify(workspaceManager, timeout(2000)).createSnapshotSync(workspace.getRuntime(), workspace.getNamespace(), workspace.getId());
         verify(runtimes, timeout(2000)).stop(any());
     }
@@ -750,7 +773,7 @@ public class WorkspaceManagerTest {
         when(workspaceMock.getStatus()).thenReturn(RUNNING);
         when(workspaceMock.getRuntime()).thenReturn(wsRuntimeMock);
         when(wsRuntimeMock.getActiveEnv()).thenReturn(testActiveEnv);
-        MachineConfigImpl machineConfig = createMachine().getConfig();
+        MachineConfigImpl machineConfig = createMachine(testWsId, testActiveEnv, false).getConfig();
 
         // when
         workspaceManager.startMachine(machineConfig, testWsId);
@@ -766,7 +789,7 @@ public class WorkspaceManagerTest {
         WorkspaceImpl workspaceMock = mock(WorkspaceImpl.class);
         doReturn(workspaceMock).when(workspaceManager).getWorkspace(testWsId);
         when(workspaceMock.getStatus()).thenReturn(STARTING);
-        MachineConfigImpl machineConfig = createMachine().getConfig();
+        MachineConfigImpl machineConfig = createMachine(testWsId, "env1", false).getConfig();
 
         // when
         workspaceManager.startMachine(machineConfig, testWsId);
@@ -774,11 +797,16 @@ public class WorkspaceManagerTest {
 
     @Test
     public void shouldBeAbleToCreateSnapshot() throws Exception {
-        // then
+        // given
         final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
         when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
         RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
         when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
 
         // when
         workspaceManager.createSnapshot(workspace.getId());
@@ -787,14 +815,12 @@ public class WorkspaceManagerTest {
         verify(workspaceManager, timeout(1_000)).createSnapshotSync(any(WorkspaceRuntimeImpl.class),
                                                     eq(workspace.getNamespace()),
                                                     eq(workspace.getId()));
-
-        // TODO add check for event
     }
 
     @Test(expectedExceptions = ConflictException.class,
           expectedExceptionsMessageRegExp = "Could not .* the workspace '.*' because its status is '.*'.")
     public void shouldNotCreateSnapshotIfWorkspaceIsNotRunning() throws Exception {
-        // then
+        // given
         final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
         when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
         RuntimeDescriptor descriptor = createDescriptor(workspace, STARTING);
@@ -806,28 +832,270 @@ public class WorkspaceManagerTest {
 
     @Test
     public void shouldSnapshotAllMachinesInWs() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(runtimes, timeout(1_000).times(2)).saveMachine(eq(workspace.getNamespace()),
+                                                              eq(workspace.getId()),
+                                                              anyString());
+        verify(snapshotDao, timeout(1_000).times(2)).saveSnapshot(any(SnapshotImpl.class));
+    }
+
+    @Test
+    public void shouldSendEventOnStartSnapshotSaving() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(eventService, timeout(1_000)).publish(eq(newDto(WorkspaceStatusEvent.class)
+                                                                .withEventType(SNAPSHOT_CREATING)
+                                                                .withWorkspaceId(workspace.getId())));
+    }
+
+    @Test
+    public void shouldSendEventOnSuccessfulSnapshotSaving() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(eventService, timeout(1_000)).publish(eq(newDto(WorkspaceStatusEvent.class)
+                                                                .withEventType(SNAPSHOT_CREATED)
+                                                                .withWorkspaceId(workspace.getId())));
+    }
+
+    @Test
+    public void shouldSendSnapshotSavingFailedEventIfDevMachineSnapshotSavingFailed() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        for (MachineImpl machine : descriptor.getRuntime().getMachines()) {
+            if (machine.getConfig().isDev()) {
+                when(runtimes.saveMachine(workspace.getNamespace(), workspace.getId(), machine.getId()))
+                        .thenThrow(new ServerException("test error"));
+            }
+        }
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(eventService, timeout(1_000)).publish(eq(newDto(WorkspaceStatusEvent.class)
+                                                                .withEventType(SNAPSHOT_CREATION_ERROR)
+                                                                .withWorkspaceId(workspace.getId())
+                                                                .withError("test error")));
+    }
+
+    @Test
+    public void shouldNotSendSnapshotSavingFailedEventIfNonDevMachineSnapshotSavingFailed() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        for (MachineImpl machine : descriptor.getRuntime().getMachines()) {
+            if (!machine.getConfig().isDev()) {
+                when(runtimes.saveMachine(workspace.getNamespace(), workspace.getId(), machine.getId()))
+                        .thenThrow(new ServerException("test error"));
+            }
+        }
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(eventService, timeout(1_000)).publish(eq(newDto(WorkspaceStatusEvent.class)
+                                                                .withEventType(SNAPSHOT_CREATED)
+                                                                .withWorkspaceId(workspace.getId())));
+    }
+
+    @Test
+    public void shouldReturnFalseOnFailureSnapshotSavingIfDevMachineSavingFailed() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        for (MachineImpl machine : descriptor.getRuntime().getMachines()) {
+            if (machine.getConfig().isDev()) {
+                when(runtimes.saveMachine(workspace.getNamespace(), workspace.getId(), machine.getId()))
+                        .thenThrow(new ServerException("test error"));
+            }
+        }
+
+        // when
+        boolean snapshotSavingStatus = workspaceManager.createSnapshotSync(
+                new WorkspaceRuntimeImpl(descriptor.getRuntime()), workspace.getNamespace(), workspace.getId());
+
+        // then
+        assertFalse(snapshotSavingStatus);
+    }
+
+    @Test
+    public void shouldReturnTrueOnSuccessfulSavingSnapshotsForSeveralMachines() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+
+        // when
+        boolean snapshotSavingStatus = workspaceManager.createSnapshotSync(
+                new WorkspaceRuntimeImpl(descriptor.getRuntime()), workspace.getNamespace(), workspace.getId());
+
+        // then
+        assertTrue(snapshotSavingStatus);
+        // ensure that multiple machines were saved
+        verify(snapshotDao, timeout(1_000).atLeast(2)).saveSnapshot(any(SnapshotImpl.class));
+    }
+
+    @Test
+    public void shouldReturnTrueOnSavingSnapshotsForSeveralMachinesWhenNonDevMachineSavingFailed() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        for (MachineImpl machine : descriptor.getRuntime().getMachines()) {
+            if (!machine.getConfig().isDev()) {
+                when(runtimes.saveMachine(workspace.getNamespace(), workspace.getId(), machine.getId()))
+                        .thenThrow(new ServerException("test error"));
+            }
+        }
+
+        // when
+        boolean snapshotSavingStatus = workspaceManager.createSnapshotSync(
+                new WorkspaceRuntimeImpl(descriptor.getRuntime()), workspace.getNamespace(), workspace.getId());
+
+        // then
+        assertTrue(snapshotSavingStatus);
     }
 
     @Test
     public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval1() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        doThrow(new NotFoundException("test error")).when(snapshotDao).removeSnapshot(anyString());
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(workspaceManager, timeout(1_000)).createSnapshotSync(any(WorkspaceRuntimeImpl.class),
+                                                                    eq(workspace.getNamespace()),
+                                                                    eq(workspace.getId()));
     }
 
     @Test
     public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval2() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        doThrow(new NotFoundException("test error")).when(runtimes).removeSnapshot(any(SnapshotImpl.class));
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(workspaceManager, timeout(1_000)).createSnapshotSync(any(WorkspaceRuntimeImpl.class),
+                                                                    eq(workspace.getNamespace()),
+                                                                    eq(workspace.getId()));
     }
 
     @Test
     public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval3() throws Exception {
-    }
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenThrow(new NotFoundException("test error"));
 
-    @Test
-    public void shouldSendSnapshotSavingFailedEventIfDevMachineSnapshotingFailed() throws Exception {
-        // TODO also check boolean return
-    }
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
 
-    @Test
-    public void shouldNotSendSnapshotSavingFailedEventIfNonDevMachineSnapshotingFailed() throws Exception {
-        // TODO also check boolean return
+        // then
+        verify(workspaceManager, timeout(1_000)).createSnapshotSync(any(WorkspaceRuntimeImpl.class),
+                                                                    eq(workspace.getNamespace()),
+                                                                    eq(workspace.getId()));
     }
 
     @Test
@@ -896,9 +1164,9 @@ public class WorkspaceManagerTest {
         assertNotNull(environment);
 
         final WorkspaceRuntimeImpl runtime = new WorkspaceRuntimeImpl(workspace.getConfig().getDefaultEnv());
-        MachineImpl machine = spy(createMachine());
+        MachineImpl machine = spy(createMachine(workspace.getId(), workspace.getConfig().getDefaultEnv(), true));
         runtime.getMachines().add(machine);
-        MachineImpl machine2 = spy(createMachine());
+        MachineImpl machine2 = spy(createMachine(workspace.getId(), workspace.getConfig().getDefaultEnv(), false));
         runtime.getMachines().add(machine2);
 
         final RuntimeDescriptor descriptor = mock(RuntimeDescriptor.class);
@@ -923,20 +1191,20 @@ public class WorkspaceManagerTest {
                                   .build();
     }
 
-    private MachineImpl createMachine() {
+    private MachineImpl createMachine(String workspaceId, String envName, boolean isDev) {
         return MachineImpl.builder()
                           .setConfig(MachineConfigImpl.builder()
-                                                      .setDev(false)
-                                                      .setName("machineName")
+                                                      .setDev(isDev)
+                                                      .setName("machineName" + UUID.randomUUID())
                                                       .setSource(new MachineSourceImpl("type").setContent("content"))
                                                       .setLimits(new MachineLimitsImpl(1024))
                                                       .setType("docker")
                                                       .build())
-                          .setId("id")
+                          .setId("id" + UUID.randomUUID())
                           .setOwner("userName")
                           .setStatus(MachineStatus.RUNNING)
-                          .setWorkspaceId("wsId")
-                          .setEnvName("envName")
+                          .setWorkspaceId(workspaceId)
+                          .setEnvName(envName)
                           .setRuntime(new MachineRuntimeInfoImpl(new HashMap<>(),
                                                                  new HashMap<>(),
                                                                  new HashMap<>()))

--- a/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
+++ b/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
@@ -1032,6 +1032,27 @@ public class WorkspaceManagerTest {
     }
 
     @Test
+    public void shouldRemoveRuntimeSnapshotIfSavingSnapshotInDaoFails() throws Exception {
+        // given
+        final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");
+        when(workspaceDao.get(workspace.getId())).thenReturn(workspace);
+        RuntimeDescriptor descriptor = createDescriptor(workspace, RUNNING);
+        when(runtimes.get(any())).thenReturn(descriptor);
+        SnapshotImpl oldSnapshot = mock(SnapshotImpl.class);
+        when(snapshotDao.getSnapshot(eq(workspace.getId()),
+                                     eq(workspace.getConfig().getDefaultEnv()),
+                                     anyString()))
+                .thenReturn(oldSnapshot);
+        doThrow(new SnapshotException("test error")).when(snapshotDao).saveSnapshot(any(SnapshotImpl.class));
+
+        // when
+        workspaceManager.createSnapshot(workspace.getId());
+
+        // then
+        verify(runtimes, timeout(1_000).times(2)).removeSnapshot(any(SnapshotImpl.class));
+    }
+
+    @Test
     public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval1() throws Exception {
         // given
         final WorkspaceImpl workspace = workspaceManager.createWorkspace(createConfig(), "user123", "account");

--- a/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
+++ b/wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/WorkspaceManagerTest.java
@@ -787,6 +787,8 @@ public class WorkspaceManagerTest {
         verify(workspaceManager, timeout(1_000)).createSnapshotSync(any(WorkspaceRuntimeImpl.class),
                                                     eq(workspace.getNamespace()),
                                                     eq(workspace.getId()));
+
+        // TODO add check for event
     }
 
     @Test(expectedExceptions = ConflictException.class,
@@ -800,6 +802,32 @@ public class WorkspaceManagerTest {
 
         // when
         workspaceManager.createSnapshot(workspace.getId());
+    }
+
+    @Test
+    public void shouldSnapshotAllMachinesInWs() throws Exception {
+    }
+
+    @Test
+    public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval1() throws Exception {
+    }
+
+    @Test
+    public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval2() throws Exception {
+    }
+
+    @Test
+    public void shouldIgnoreNotFoundExceptionOnOldSnapshotRemoval3() throws Exception {
+    }
+
+    @Test
+    public void shouldSendSnapshotSavingFailedEventIfDevMachineSnapshotingFailed() throws Exception {
+        // TODO also check boolean return
+    }
+
+    @Test
+    public void shouldNotSendSnapshotSavingFailedEventIfNonDevMachineSnapshotingFailed() throws Exception {
+        // TODO also check boolean return
     }
 
     @Test

